src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -646,14 +646,22 @@ public void AddDynamicSourceFile(string dynamicFilePath, ImmutableArray<string>
 
                 fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
 
-                // remember map between original dynamic file path to DynamicFileInfo.FilePath
-                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
-                _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
-                return;
+                lock (_gate)
+                {
+
+                    // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                    _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
```

Comment:
what happens if two threads eventually get here.  this will throw, right?

Reply:
Yes; this is made explicit now with a proper exception message.